src/UglyToad.PdfPig/Writer/PdfMerger.cs

```diff
@@ -214,30 +263,41 @@ public void Close()
                 context.Dispose();
             }
 
-            private (IndirectReferenceToken, int) CopyPagesTree(PageTreeNode treeNode, IndirectReferenceToken treeParentReference, IPdfTokenScanner tokenScanner)
+            private (IndirectReferenceToken, int) CopyPagesTree(PageTreeNode treeNode, IndirectReferenceToken treeParentReference, IPdfTokenScanner tokenScanner, Func<bool> canCopyPage)
             {
                 Debug.Assert(!treeNode.IsPage);
-                
+
                 var currentNodeReference = context.ReserveNumberToken();
 
                 var pageReferences = new List<IndirectReferenceToken>();
                 var nodeCount = 0;
                 foreach (var pageNode in treeNode.Children)
                 {
-                    IndirectReferenceToken newEntry;
+                    IndirectReferenceToken newEntry = null;
                     if (!pageNode.IsPage)
                     {
                         var count = 0;
-                        (newEntry, count) = CopyPagesTree(pageNode, currentNodeReference, tokenScanner);
+                        (newEntry, count) = CopyPagesTree(pageNode, currentNodeReference, tokenScanner, canCopyPage);
                         nodeCount += count;
                     }
                     else
                     {
-                        newEntry = CopyPageNode(pageNode, currentNodeReference, tokenScanner);
-                        ++nodeCount;
+                        if (canCopyPage())
```

Comment:
Is`pageNode.PageNumber` reliable here ? If so, I would have expected `canCopyPage` to be an `ISet<int> selectedPages`. This could also lead, by changing the loop logic line 225/274 https://github.com/UglyToad/PdfPig/pull/245/files?diff=unified&w=1#diff-494177227c2134c490310f8e8141d5033956b4872afedaf2a9cfc3d30fe27816R274 to allow pages reordering.

Reply:
You are totally right, I should use an `ISet`

The problem with copying pages out of order arise when we take in considerations the resources that those pages share with each others. 

Example

This is my main document
- Root (Page Tree)
  - Tree (Child Page Tree)
    - Resources
      - Image (F1) (Logo)
      - Font (F2) (Custom Font)
    - Page (Page No. 1)
      - Resources
        - Image (F3) (User Icon)
    - Page (Page No. 2)
      - Resources
        - Image (F3) (User Icon)
  - Tree (Child Page Tree)
    - Resources
      - Image (F1) (Another Logo)
      - Font (F2) ( Another Custom Font)
    - Page (Page No. 3)
      - Resources
        - Image (F3) (User Icon)


Now, let's try to merge the selection of pages (1, 3, 2) in that order:

- Root (Page Tree)
  - Tree (Child Page Tree)
    - Resources
      - Image (F1) (Logo)
      - Font (F2) (Custom Font)
    - Page (Page No. 1)
      - Resources
        - Image (F3) (User Icon)
  - Tree (Child Page Tree)
    - Resources
      - Image (F1) (Another Logo)
      - Font (F2) ( Another Custom Font)
    - Page (Page No. 3)
      - Resources
        - Image (F3) (User Icon)
  - Tree (Child Page Tree)
    - Resources
      - Image (F1) (Logo)
      - Font (F2) (Custom Font)
    - Page (Page No. 2)
      - Resources
        - Image (F3) (User Icon)

As you can see we end up with duplicated resources and resources name could collide. This is no problem with small files. But file with many pages (>1000) that share many resources, we would bloat the file and the computer resources.

My idea for solving this was to copy pages from page to tree (bottom to top) instead of tree to page (top to bottom). That way we can walk up the resources, making sure none of them collide and reusing the `IndirectReference` when the resource is reuse. But, this approach is a little more difficult to implement.